### DISCUSSION TITLE
Multiple flask apps, unable to communicate with socketio

### DISCUSSION THREAD

**Discussion by weevaa** [miguelgrinberg/Flask-SocketIO#1768] [Q&A]
Hey Miguel!

I have two Flask apps running on the same machine in separate venv's using supervisord, nginx and gunicorn

The main app handles front-end (port 8000) and works perfectly with socketio.
The other handles bulky/longer running automation scripts. (port 3000) this one I am not seeing emit messages

I want the secondary app/instance to emit process updates to the user... 

In an attempt to solve my problem I came across: [This past issue.](https://github.com/miguelgrinberg/Flask-SocketIO/issues/643)  I believe I have it all setup properly, however the secondary app (bgsvc) show's it is emitting but I'm not seeing any of those messages on the front end.... and when emitting a test message from the front, it does not appear to see that message. 

NGINX setup:
```
upstream flip_nodes {
    # Socket.IO requires sticky sessions
    ip_hash;
    server 127.0.0.1:8000; 
    server 127.0.0.1:3000;
    # to scale the app, just add more nodes here! USE IP ADDRESS NOT HOSTNAME
}

 location /socket.io {
        proxy_pass http://flip_nodes/socket.io;
        proxy_http_version 1.1;
        proxy_redirect off;
        proxy_buffering off;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
    }

location / {
        proxy_pass http://flip_nodes;
        proxy_redirect off;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    }

location /bulk {
       proxy_pass http://127.0.0.1:3000/bulk;
    }
```

the /bulk part works as intended (just emit events are not seen by the front end)

SupervisorD config app1: 
`command=/***/main/bin/gunicorn -k geventwebsocket.gunicorn.workers.GeventWebSocketWorker -w 1 main:app`

app2:
`command=/***/bgsvc/bin/gunicorn -k geventwebsocket.gunicorn.workers.GeventWebSocketWorker -w 1 main:app --bind 0.0.0.0:3000`

both apps have same cors_allowed_origins including FQDN and are started like: 
```
socketcors = ['http://127.0.0.1:8000','http://127.0.0.1:3000','http://127.0.0.1:5000']
socketio = SocketIO(app, cors_allowed_origins=socketcors)

if __name__ == '__main__':
    socketio.run(app, host="0.0.0.0", port=5000)
```
I tried starting them with port as 8000 / 3000 (respectively), both on same port, as well as with just: 
`socketio.run(app)`

any of the other ways I start them, the main app sends / recieves all emit events but the bgsvc does not.  
 
Socketio debugging (for bgsvc) shows:
```
[2022-01-09 04:25:22 -0500] [19354] [INFO] Starting gunicorn 20.1.0
[2022-01-09 04:25:22 -0500] [19354] [INFO] Listening at: http://0.0.0.0:3000 (19354)
[2022-01-09 04:25:22 -0500] [19354] [INFO] Using worker: geventwebsocket.gunicorn.workers.GeventWebSocketWorker
[2022-01-09 04:25:22 -0500] [19357] [INFO] Booting worker with pid: 19357
Server initialized for gevent.
emitting event "ProcessUpdate" to 5 [/]
emitting event "ProcessUpdate" to 5 [/]
emitting event "ProcessUpdate" to 5 [/]
emitting event "ProcessUpdate" to 5 [/]
```

for all events called with: 
```
def process_update(userObj, msg):
    print(msg)
    socketio.emit("ProcessUpdate",msg,room=str(userObj.id))
```

both virtual environments are running: 
```
Flask-SocketIO==5.1.1
python-engineio==4.2.0
python-socketio==5.3.0
gevent==21.12.0
gevent-websocket==0.10.1
greenlet==1.1.2
gunicorn==20.1.0
```

Hopefully there is something that I can do to get this working.  Otherwise I'm thinking create a route in the main app, that the secondary app makes a request to that will forward the message I want, but figured I'm missing something and you could help point me in the right direction.  

Thank you in advance!

**Comment by miguelgrinberg** [comment]
Looks like you haven't added a message queue for these two apps to communicate with each other: https://flask-socketio.readthedocs.io/en/latest/deployment.html#emitting-from-an-external-process.
